### DISCUSSION TITLE
Can't resolve 'env' in .../pkg - how to investigate?

### DISCUSSION THREAD

**Discussion by kalzoo** [wasm-bindgen/wasm-bindgen#3500] [Q&A] [answered]
Following on from issue reported [here](https://github.com/rustwasm/wasm-bindgen/issues/1907) and closed.

My situation in particular:

* private crate, private source
* `wasm-bindgen` has been working smoothly for about a year across many releases
* WASM package is then consumed by a React frontend. The error appears when that frontend is built, in release or dev mode.
* In the course of one particular, small-scope MR, the task began to fail with the titular error:

```
Module not found: Error: Can't resolve 'env' in '<redacted>/pkg'
```

* No crates or JS deps have been updated in `Cargo.toml` or `package.json`
* I can reproduce locally and in CI, not a caching/environment issue
* Building the package with `wasm-pack build`, both `--debug` and not

## What I've done

* Use `wasm2wat` to text-ify the output. That comes out to roughly 7.6M LoC with no clear indications of the dangling reference
* Created an empty `env` package and linked that into the package with the following. **This at least makes the JS build work.** So the missing function reference is only one level deep; not sure how much of a revelation that is.

## So, my question - what's the right way to investigate this?

* It _appears_ that something about one of the Rust dependencies changed so that something which _was_ able to be built with WASM can no longer be
* It's not clear to me how that could be the case and why that's not a compilation failure
* It's also not clear how I can go about discovering which external reference in the WASM package no longer exists
* I can diff the (working package versions) with the (no longer working) from CI logs (the lockfile is not source-controlled), but even with a short list of packages, unless something is called out in that package's github issues etc I'm not sure the culprit will be obvious. And - even if it is obvious in my case - this discussion may be useful for others later.

**Comment by Liamolucko** [comment, marked as answer]
You can pretty easily see the name of the import that's causing an issue using `wasm2wat` by searching for `(import "env"`, which should show you one or more lines like this:

```wasm
(import "env" "function" (func $function (type 4)))
```

Unfortunately though all the information about what crate the import came from is erased.

However, `cargo` outputs a separate object file for every crate you depend on in `target/wasm32-unknown-unknown/debug/deps` (wrapped up inside an `rlib`), so you could figure out which crate caused the offending import by looking through all of those to see which one imports it.

I got kind of over-invested in this and created a tool which does that automatically: https://github.com/Liamolucko/find_wasm_import. You have to use `wasm2wat` to find what the name of one of the offending imports is first, but after that you just point it at the `deps` folder and it should tell you the crate which is causing the problem:

```
> find_wasm_import env function target/wasm32-unknown-unknown/debug/deps
"env" "function" is imported by:
  target/wasm32-unknown-unknown/debug/deps/liboffending_crate-037c8cef292eb070.rlib
```

Also, by the way, the error isn't necessarily cropping up now because of something changing about your dependencies. Imported functions like the one causing this error only get included in the output wasm if they're used, so it might be happening because you started using some new feature of a dependency you already had which relies on some wasm-incompatible import.

**Reply by daxpedda** [reply]
> I got kind of over-invested in this and created a tool which does that automatically: https://github.com/Liamolucko/find_wasm_import.

That's pretty sick @Liamolucko, thank you! That's a very useful tool.

**Reply by kalzoo** [reply]
Thanks, @Liamolucko , that's awesome! Indeed, this came down to the `instant` [crate](https://github.com/sebcrozet/instant), which joined the dependency tree as a dep-of-a-dep-of-a-dep and with which I wasn't yet familiar. Curiously enough, it specifically exists for wasm, but I had to enable its `wasm-bindgen` feature. Back in business! Would have been hard to solve this without this tool you wrote. Worth adding a note & link in the docs and maybe pulling it into the repo itself?

**Reply by frewsxcv** [reply]
Thank you for this thread and utility. In my case, it is was `instant`:

```
% find_wasm_import env now target/wasm32-unknown-unknown/release/deps
"env" "now" is imported by:
  target/wasm32-unknown-unknown/release/deps/libinstant-84b16be68a676ab1.rlib
```

**Reply by xangelix** [reply]
@Liamolucko you saved my life! Really disappointed there isn't any error sooner than runtime. Do you know of any progress that has been made on that front, to error at compile time?

**Reply by damccull** [reply]
I am having this imported 'env' issue under a dioxus app, which in turn uses wasm-bindgen. I don't have the characters 'env' anywhere in my wasm2wat code though. How can I find the offender without that?

**Reply by kossnocorp** [reply]
@Liamolucko thank you a lot!

Here are some notes from me for others struggling with the `env` problem, explaining step by step how to use `find_wasm_import`:

First, mise config:

```toml
# ...
"npm:wabt" = "latest"
"cargo:https://github.com/Liamolucko/find_wasm_import" = { version = "rev:b726e870fcc30f59987ef59ad448ceb2d3dd998d", bin = "find_wasm_import" }
# ...
```

Then convert using `wasm2wat`:

```bash
wasm2wat pkgs/npm/volumen_bg.wasm > tmp/volumen.wat
```

Find the lines (`mise use rg`):

```
rg -e 'import "env"' tmp/volumen.wat
#=> 86:  (import "env" "iswlower" (func (;5;) (type 3)))
#=> 87:  (import "env" "memchr" (func (;6;) (type 4)))
#=> 88:  (import "env" "iswupper" (func (;7;) (type 3)))
#=> 89:  (import "env" "iswxdigit" (func (;8;) (type 3)))
```

And then finally use `find_wasm_import`:

```bash
$ find_wasm_import env iswxdigit target/wasm32-unknown-unknown/debug/deps
#=> "env" "iswxdigit" is imported by:
#=>   target/wasm32-unknown-unknown/debug/deps/libtree_sitter_php-0566b6314d309efc.rlib
#=>   target/wasm32-unknown-unknown/debug/deps/libtree_sitter_php-2a2a6e4bf8d40cd7.rlib
```

In my case, that was Tree-sitter crates `tree-sitter-php` and `tree-sitter-ruby` that used `wctype.h` and required shims.

**Comment by yunho-c** [comment]
[I encountered a similar error in `pykeio/ort` cate](https://github.com/pykeio/ort/issues/75) and resolved by switching `rustc` to nightly branch via rustup.
